core/server/common/src/main/java/alluxio/util/ParallelZipUtils.java

```diff
@@ -0,0 +1,181 @@
+              return new NullInputStream(0);
+            }
+          };
+
+          String entryName = dirPath.relativize(subPath).toString();
+          if (file.isDirectory()) {
+            entryName += File.separator;
+          }
+
+          ZipArchiveEntry zipArchiveEntry = new ZipArchiveEntry(entryName);
+          zipArchiveEntry.setMethod(ZipArchiveEntry.DEFLATED);
+
+          parallelScatterZipCreator.addArchiveEntry(zipArchiveEntry, inputStreamSupplier);
+        }
+      }
+
+      parallelScatterZipCreator.writeTo(zipArchiveOutputStream);
+      zipArchiveOutputStream.flush();
+      fileOutputStream.getFD().sync();
+    } catch (ExecutionException e) {
+      LOG.error("Parallel compress rocksdb fail", e);
+      ExecutorServiceUtils.shutdownAndAwaitTermination(executor);
+      FileUtils.deletePathRecursively(backupPath);
+      throw new RuntimeException(e);
+    }
+
+    LOG.info("compress in parallel from path " + dirPath + " to " + backupPath
+        + " statistics message "
+        + parallelScatterZipCreator.getStatisticsMessage().toString());
+  }
+
+  /**
+   * Reads a zipped archive from a path in parallel and writes it to the given path.
+   *
+   * @param dirPath
+   * @param backupPath
+   * @param poolSize
+   */
+  public static void deCompress(Path dirPath, String backupPath, int poolSize) throws IOException {
+    LOG.info("deCompress in parallel from path " + backupPath + " to " + dirPath);
+    ExecutorService executor = ExecutorServiceFactories.fixedThreadPool(
+        "parallel-zip-deCompress-pool", poolSize).create();
+    CompletionService<Boolean> completionService = new ExecutorCompletionService<Boolean>(executor);
+
+    try (ZipFile zipFile = new ZipFile(backupPath)) {
+      Enumeration<ZipArchiveEntry> entries = zipFile.getEntries();
+      int taskCount = 0;
+
+      while (entries.hasMoreElements()) {
+        taskCount += 1;
+        ZipArchiveEntry entry = entries.nextElement();
+        completionService.submit(() -> {
+          unzipEntry(zipFile, entry, dirPath);
+          return true;
+        });
+      }
+
+      for (int i = 0; i < taskCount; i++) {
+        completionService.take().get();
+      }
+
+      ExecutorServiceUtils.shutdownAndAwaitTermination(executor);
+    } catch (Exception e) {
+      LOG.error("Parallel deCompress rocksdb fail", e);
+      ExecutorServiceUtils.shutdownAndAwaitTermination(executor);
+      FileUtils.deletePathRecursively(dirPath.toString());
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Unzip entry in ZipFile.
+   *
+   * @param zipFile
+   * @param entry
+   * @param dirPath
+   * @throws Exception
+   */
+  private static void unzipEntry(ZipFile zipFile, ZipArchiveEntry entry, Path dirPath)
+      throws Exception {
+    File outputFile = new File(dirPath.toFile(), entry.getName());
+    outputFile.getParentFile().mkdirs();
+
+    if (entry.isDirectory()) {
+      outputFile.mkdir();
```

Comment:
output to target file under the directory?

Reply:
yes

Reply:
The first entry should be the directory itself. This util supports compression and decompression with nested directories